### PR TITLE
Implement Docker client

### PR DESCRIPTION
**Pull request by rizul2108**
resolves issue  "Use docker client to build images" #63

### CODE REVIEW

File: lib/utils/docker.go [hunk 1]
```diff
@@ -1,38 +1,111 @@
...
+	if err != nil {
+		log.Fatal(err, " :unable to push docker image")
+		return
 	}
-	log.Println("Docker image pushed successfully")
+	defer pushResponse.Close()
+
 }
 
 func DockerLogin(username string, password string) {
```
Comment (Bisht13) [inline lib/utils/docker.go:72]: Should be the first function

File: lib/utils/docker.go [hunk 2]
```diff
@@ -1,38 +1,111 @@
...
-	log.Println("Building docker image, Please wait...")
-	cmd := "docker build -t " + "harbor.katana.local/katana/" + _ChallengeName + " " + _DockerfilePath
-	if err := RunCommand(cmd); err != nil {
-		log.Printf("Error: %s\n", err)
+	buf := new(bytes.Buffer)
+	if err := Tar(_DockerfilePath, buf); err != nil {
+		log.Fatal(err, ": error tarring directory")
+		return
+	}
+	ctx := context.Background()
```
Comment (Bisht13) [inline lib/utils/docker.go:21]: Don't need to create a variable for context, call it directly where required

File: lib/utils/docker.go [hunk 3]
```diff
@@ -1,38 +1,111 @@
...
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		log.Fatal(err)
+		return
+	}
+	log.Println("Building Docker image, Please wait......")
```
Comment (Bisht13) [inline lib/utils/docker.go:27]: Give a line break before and after

File: lib/utils/docker.go [hunk 4]
```diff
@@ -1,38 +1,111 @@
...
+		},
+	)
+	if err != nil {
+		log.Fatal(err, " :unable to create image")
+		return
 	}
+
+	defer imageBuildResponse.Body.Close()
 	log.Println("Docker image built successfully")
+	DockerLogin(configs.KatanaConfig.Harbor.Username, configs.KatanaConfig.Harbor.Password)
```
Comment (Bisht13) [inline lib/utils/docker.go:44]: Is this required?

File: lib/utils/docker.go [hunk 5]
```diff
@@ -1,38 +1,111 @@
...
+
 }
 
 func DockerLogin(username string, password string) {
 	log.Println("Logging into Harbor, Please wait...")
-	cmd := "docker login -u " + username + " -p " + password + " " + "harbor.katana.local"
-	if err := RunCommand(cmd); err != nil {
-		log.Printf("Error: %s\n", err)
+
+	ctx := context.Background()
```
Comment (Bisht13) [inline lib/utils/docker.go:75]: Variable not needed

File: lib/utils/docker.go [hunk 6]
```diff
@@ -1,38 +1,111 @@
...
+		return
+	}
+
 	log.Println("Logged into Harbor successfully")
 }
 
 func DockerImageExists(imageName string) bool {
-	cmd := "docker image inspect " + "harbor.katana.local/katana/" + imageName
-	if err := RunCommand(cmd); err != nil {
+	ctx := context.Background()
```
Comment (Bisht13) [inline lib/utils/docker.go:102]: Variable not needed

**Review by Bisht13 — CHANGES_REQUESTED** [review]

File: lib/utils/docker.go (hunk 3 above)
Comment (rizul2108) [reply lib/utils/docker.go:27]: while printing this statement or in the code only ?

Comment (Bisht13) [reply lib/utils/docker.go:27]: Here, as well as other places you seem fit, it becomes more readable that way.

File: lib/utils/docker.go (hunk 4 above)
Comment (rizul2108) [reply lib/utils/docker.go:44]: yes without it , not pushing image to harbor

**Pull request merged**